import sys
# åˆ‡æ¢åˆ° PyQt6 åº“
from PyQt6.QtWidgets import (
    QApplication, QMainWindow, QWidget,
    QVBoxLayout, QHBoxLayout, QLabel,
    QLineEdit, QPushButton, QSlider,
    QScrollArea
)
# å¯¼å…¥ Qt æ ¸å¿ƒç»„ä»¶å’Œæšä¸¾å€¼
from PyQt6.QtCore import Qt, QTimer, QSize, pyqtSignal as Signal, QObject, QThreadPool
from PyQt6.QtGui import QFont, QIcon, QAction

# å¯¼å…¥ gRPC Worker å’Œ proto æ¶ˆæ¯ (ç¡®ä¿è¿™äº›æ–‡ä»¶åœ¨æ­£ç¡®çš„ä½ç½®)
from gRPCWorker import PromptWorker, WorkerSignals, StreamWorker
import proto.flowradio_pb2 as pb
from proto import flowradio_pb2_grpc as pb_grpc # ä»…åœ¨éœ€è¦æ—¶
# --- 1. ä¸»çª—å£ç±»å®šä¹‰ ---
class FlowRadioApp(QMainWindow):
    
    # çŠ¶æ€ä¿¡å·ï¼Œç”¨äºæ¥æ”¶å®æ—¶æµæ›´æ–°
    stream_update_signal = Signal(object) 
    
    def __init__(self):
        super().__init__()
        self.setWindowTitle("FlowRadio - æ‹ŸäººåŒ–æ™ºèƒ½ç”µå°")
        self.setFixedSize(600, 800) 
        
        # æ ¸å¿ƒï¼šçŠ¶æ€ç®¡ç† (ç®€åŒ–ä¸º Go åç«¯ç®¡ç†å†å²ï¼Œå‰ç«¯åªå­˜çŠ¶æ€)
        self.host_state = {
            'current_genre': 'lofi',    
            'current_memory': '',       # å­˜å‚¨ LLM è¿”å›çš„æœ€æ–° memory æ‘˜è¦
        }

        # åŠ è½½ QSS æ ·å¼ (å‡è®¾ QSS æ–‡ä»¶åœ¨ qss/ ç›®å½•ä¸‹)
        self.switch_theme('ios')

        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        
        main_layout = QVBoxLayout(central_widget)
        main_layout.setContentsMargins(20, 20, 20, 20)
        main_layout.setSpacing(15)
        
        # ä¾æ¬¡æ·»åŠ é¡¶éƒ¨ã€ä¸­éƒ¨ã€åº•éƒ¨åŒºåŸŸ
        self.top_widget = self._create_top_bar()
        self.middle_widget = self._create_middle_section()
        self.bottom_widget = self._create_bottom_bar()
        
        main_layout.addWidget(self.top_widget)
        main_layout.addWidget(self.middle_widget, 4) # æƒé‡ä¸º4ï¼Œå æ®æ›´å¤šç©ºé—´
        main_layout.addWidget(self.bottom_widget, 1) # æƒé‡ä¸º1
        
        # è¿æ¥ä¿¡å·ä¸æ§½
        self._connect_signals()
        
        # åˆå§‹åŒ–çº¿ç¨‹æ± 
        self.threadpool = QThreadPool.globalInstance()
        print(f"ThreadPool åˆå§‹åŒ–ï¼Œæœ€å¤§çº¿ç¨‹æ•°: {self.threadpool.maxThreadCount()}")
        
        # å¯åŠ¨ Stream Worker (å®æ—¶ç›‘å¬ Go åç«¯æ¨é€)
        self._start_stream_worker()


    # --- 2. é¡¶éƒ¨åŒºåŸŸï¼šDJ & å¿«æ·é”® ---
    def _create_top_bar(self):
        top_bar = QWidget()
        top_layout = QHBoxLayout(top_bar)
        top_layout.setContentsMargins(0, 0, 0, 0)
        
        # 1. ä¸»æŒäººä¿¡æ¯ (å·¦ä¾§)
        dj_info_widget = QWidget()
        dj_info_layout = QVBoxLayout(dj_info_widget)
        
        self.dj_name_label = QLabel(" DJ Astro")
        self.dj_status_label = QLabel("Status: ğŸŒŒ å‡†å¤‡å°±ç»ª")
        
        self.dj_name_label.setObjectName("DjNameLabel")
        self.dj_status_label.setObjectName("DjStatusLabel")
        
        dj_info_layout.addWidget(self.dj_name_label)
        dj_info_layout.addWidget(self.dj_status_label)
        
        top_layout.addWidget(dj_info_widget)
        top_layout.addStretch(1) # å¼¹æ€§ç©ºé—´
        
        # 2. é£æ ¼åˆ‡æ¢å¿«æ·é”® (å³ä¾§)
        style_shortcuts_widget = QWidget()
        shortcuts_layout = QHBoxLayout(style_shortcuts_widget)
        shortcuts_layout.setSpacing(5)

        # ç¤ºä¾‹å¿«æ·é”®æŒ‰é’®
        self.btn_style_lofi = QPushButton("Lo-Fi")
        self.btn_style_ambient = QPushButton("Ambient")
        
        shortcuts_layout.addWidget(self.btn_style_lofi)
        shortcuts_layout.addWidget(self.btn_style_ambient)
        
        top_layout.addWidget(style_shortcuts_widget)
        
        return top_bar

    # --- 3. ä¸­éƒ¨åŒºåŸŸï¼šåŠ¨æ€ä¿¡æ¯æµä¸æ§åˆ¶ ---
    def _create_middle_section(self):
        middle_section = QWidget()
        main_layout = QVBoxLayout(middle_section)
        main_layout.setContentsMargins(0, 0, 0, 0)

        # 1. æ’­æ”¾æ§åˆ¶æ¡ (é¡¶éƒ¨)
        controls_widget = QWidget()
        controls_layout = QHBoxLayout(controls_widget)
        controls_layout.setContentsMargins(0, 0, 0, 0)
        
        self.btn_play_pause = QPushButton("â¸ï¸")
        self.btn_play_pause.setFixedSize(40, 40)
        
        self.volume_slider = QSlider(Qt.Orientation.Horizontal) 
        self.volume_slider.setRange(0, 100)
        self.volume_slider.setValue(50)
        
        controls_layout.addWidget(self.btn_play_pause)
        controls_layout.addWidget(QLabel("Vol:"))
        controls_layout.addWidget(self.volume_slider)
        
        main_layout.addWidget(controls_widget)
        
        # 2. ç•™è¨€æ»šåŠ¨åŒº/åŠ¨æ€æ–‡å­—æ˜¾ç¤ºåŒº (æ ¸å¿ƒ)
        self.message_area = QScrollArea()
        self.message_area.setObjectName("MessageScrollArea")
        self.message_area.setWidgetResizable(True)
        
        self.message_area.setHorizontalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)

        # QScrollAreaéœ€è¦ä¸€ä¸ªå†…å®¹Widget
        self.message_content_widget = QWidget()
        self.message_content_widget.setObjectName("MessageContentWidget")
        self.message_layout = QVBoxLayout(self.message_content_widget)
        self.message_layout.setSpacing(8)
        
        self.message_layout.setAlignment(Qt.AlignmentFlag.AlignTop) 

        self.message_area.setWidget(self.message_content_widget)
        
        main_layout.addWidget(self.message_area)
        
        return middle_section
    
    # --- 4. åº•éƒ¨åŒºåŸŸï¼šå¬ä¼—æ¥ç”µä¸é…ç½® ---
    def _create_bottom_bar(self):
        bottom_bar = QWidget()
        bottom_bar.setObjectName("BottomBarWidget")
        bottom_layout = QHBoxLayout(bottom_bar)
        
        # 1. é…ç½®æŒ‰é’® (å·¦ä¾§)
        self.btn_config = QPushButton("âš™ï¸ é…ç½®")
        self.btn_config.setObjectName("ConfigButton")
        bottom_layout.addWidget(self.btn_config)
        
        # 2. ç•™è¨€æ¡† (ä¸­é—´)
        self.input_prompt = QLineEdit()
        self.input_prompt.setPlaceholderText("è¾“å…¥æ‚¨çš„éœ€æ±‚ï¼Œå¦‚ï¼šæ”¾ç‚¹é€‚åˆå†™ä»£ç çš„éŸ³ä¹...")
        self.input_prompt.setObjectName("PromptInput")
        bottom_layout.addWidget(self.input_prompt)
        
        # 3. æ‹¨æ‰“æŒ‰é’® (å³ä¾§)
        self.btn_call_in = QPushButton("ğŸ“ CALL IN")
        self.btn_call_in.setObjectName("CallInButton")
        self.btn_call_in.setFixedSize(120, 35)
        bottom_layout.addWidget(self.btn_call_in)
        
        return bottom_bar

    # --- 5. ä¿¡å·ä¸æ§½è¿æ¥ ---
    def _connect_signals(self):
        # æ‹¨æ‰“æŒ‰é’®ç‚¹å‡»äº‹ä»¶
        self.btn_call_in.clicked.connect(self._handle_call_in)
        # è¾“å…¥æ¡†å›è½¦äº‹ä»¶ (æŒ‰ä¸‹å›è½¦ä¹Ÿè§†ä¸ºæ‹¨æ‰“)
        self.input_prompt.returnPressed.connect(self._handle_call_in)
        # æ’­æ”¾/æš‚åœæŒ‰é’®ç‚¹å‡»äº‹ä»¶
        self.btn_play_pause.clicked.connect(self._handle_play_pause)
        
        # æ ·å¼æŒ‰é’®ç‚¹å‡»ç¤ºä¾‹
        self.btn_style_lofi.clicked.connect(lambda: self.add_message("ç³»ç»Ÿï¼šåˆ‡æ¢è‡³ Lo-Fi é£æ ¼"))
        self.btn_style_lofi.clicked.connect(lambda: self.switch_theme('synthwave'))
        # Ambient æŒ‰é’®åˆ‡æ¢å› iOS é£æ ¼
        self.btn_style_ambient.clicked.connect(lambda: self.switch_theme('ios'))

    def _handle_call_in(self):
        prompt_text = self.input_prompt.text().strip()
        if prompt_text:
            
            # 1. UIåé¦ˆï¼šæ˜¾ç¤ºæ¶ˆæ¯ï¼Œæ¸…ç©ºè¾“å…¥æ¡†
            self.add_message(f"æ‚¨ï¼š{prompt_text}", is_user=True)
            self.input_prompt.clear()
            
            # 2. é”å®šUIï¼Œé¿å…é‡å¤å‘é€
            self.btn_call_in.setEnabled(False)
            self.btn_call_in.setText("è¿çº¿ä¸­...")

            # 3. åˆ›å»º Worker å¹¶è¿æ¥ä¿¡å·
            # å‡è®¾å½“å‰ä¸Šä¸‹æ–‡æ˜¯ "Coding" (Visionæ¨¡å—æœªå®ç°æ—¶çš„å ä½ç¬¦)
            context_scene = "Coding" 
            
            # PromptWorker ç°åœ¨åªéœ€å‘é€æœ¬æ¬¡è¾“å…¥
            worker = PromptWorker(prompt=prompt_text, context=context_scene) 
            
            # è¿æ¥ Worker ä¿¡å·åˆ° UI çš„ Slot
            worker.signals.prompt_sent.connect(self._handle_prompt_sent)
            worker.signals.error.connect(self._handle_worker_error)
            # Worker ç»“æŸä¸ä»£è¡¨ LLM å®Œæˆï¼Œæ•…ä¸è¿æ¥ finished åˆ° unlock

            # 4. å¯åŠ¨ Worker
            self.threadpool.start(worker)
            
    # --- æ–°å¢ Stream Worker å¯åŠ¨å’Œå¤„ç†é€»è¾‘ ---
    def _start_stream_worker(self):
        """å¯åŠ¨åå°çº¿ç¨‹ï¼ŒæŒç»­ç›‘å¬ Go åç«¯æ¨é€çš„å®æ—¶æ›´æ–°"""
        worker = StreamWorker()
        
        # è¿æ¥ Worker çš„ update_received ä¿¡å·åˆ° UI çš„å¤„ç†æ§½
        worker.signals.update_received.connect(self._handle_stream_update) 
        worker.signals.error.connect(self._handle_worker_error)
        
        self.threadpool.start(worker)

    def _handle_stream_update(self, update_message: pb.UpdateMessage):
        """å¤„ç† Go åç«¯æ¨é€æ¥çš„ UpdateMessage å®æ—¶æ•°æ®"""
        
        update_type = update_message.type
        
        if update_type == pb.UpdateMessage.DJ_DECISION:
            # è§£æå†³ç­–è´Ÿè½½
            decision = update_message.decision_data 
            
            primary_prompt = decision.music_prompts[0] if decision.music_prompts else self.host_state['current_genre']
            
            # 1. æ›´æ–° UI è„šæœ¬
            self.add_message(decision.dj_script, is_user=False)
            self.dj_status_label.setText(f"Status: ğŸ¶ {primary_prompt} (ç†ç”±: {decision.action_reason})")
            
            # 2. TODO: æ’­æ”¾éŸ³é¢‘ (ä½¿ç”¨ mpv æ’­æ”¾ decision.audio_data_bytes)
            
            # 3. æ›´æ–°æœ¬åœ°çŠ¶æ€
            self.host_state['current_memory'] = decision.new_conversation_memory
            if decision.music_prompts and decision.music_prompts[0] != self.host_state['current_genre']:
                 self.host_state['current_genre'] = decision.music_prompts[0]
            
            # LLM æµç¨‹å®Œæˆï¼Œè§£é”æŒ‰é’®
            self._unlock_call_in()

        elif update_type == pb.UpdateMessage.VIRTUAL_COMMENT:
            self.add_message(update_message.virtual_comment_text, is_user=False)
        
        elif update_type == pb.UpdateMessage.SYSTEM_STATUS:
            self._handle_worker_error(update_message.system_status_data.message)

    def _handle_prompt_sent(self, success: bool):
        """ å¤„ç† Prompt è¯·æ±‚å‘é€åçš„ Go åç«¯ç¡®è®¤ä¿¡æ¯ """
        if success:
            # ä»…æ˜¾ç¤ºçŠ¶æ€ï¼Œç­‰å¾… StreamWorker æ¨é€æœ€ç»ˆç»“æœ
            self.dj_status_label.setText("Status: ğŸ§ DJ Brain æ­£åœ¨å¤„ç†...") 
        else:
            self.dj_status_label.setText("Status: âŒ Go åç«¯è¯·æ±‚å¤±è´¥")
            self._unlock_call_in() # è¯·æ±‚å¤±è´¥ï¼Œç«‹å³è§£é”

    def _handle_worker_error(self, error_message: str):
        """ å¤„ç† gRPC é€šä¿¡é”™è¯¯æˆ– StreamWorker é”™è¯¯ """
        self.add_message(f"ç³»ç»Ÿé”™è¯¯: {error_message}", is_user=False)
        self.dj_status_label.setText("Status: âŒ é€šä¿¡é”™è¯¯")
        self._unlock_call_in()

    def _unlock_call_in(self):
        """ æ— è®ºæˆåŠŸæˆ–å¤±è´¥ï¼Œéƒ½åœ¨ Worker ç»“æŸåè§£é”æŒ‰é’® """
        self.btn_call_in.setEnabled(True)
        self.btn_call_in.setText("ğŸ“ CALL IN")
        
    def _handle_play_pause(self):
        if self.btn_play_pause.text() == "â¸ï¸":
            self.btn_play_pause.setText("â–¶ï¸")
            self.dj_status_label.setText("Status: â¸ï¸ æš‚åœä¸­")
            # TODO: è°ƒç”¨åç«¯APIæš‚åœéŸ³ä¹
        else:
            self.btn_play_pause.setText("â¸ï¸")
            self.dj_status_label.setText("Status: ğŸ¶ æ­£åœ¨æ’­æ”¾")
            # TODO: è°ƒç”¨åç«¯APIæ’­æ”¾éŸ³ä¹

    # --- 6. æ ¸å¿ƒåŠŸèƒ½ï¼šåŠ¨æ€æ·»åŠ æ¶ˆæ¯ ---
    def add_message(self, text, is_user=False):
        """åŠ¨æ€æ·»åŠ ä¸€æ¡æ¶ˆæ¯åˆ°æ»šåŠ¨åŒº"""
        msg_label = QLabel(text)
        msg_label.setWordWrap(True)
        msg_label.setObjectName("UserMessage" if is_user else "SystemMessage")

        # æ’å…¥æ–°æ¶ˆæ¯
        self.message_layout.addWidget(msg_label)

        # ç¡®ä¿æ»šåŠ¨æ¡è‡ªåŠ¨æ»šåŠ¨åˆ°åº•éƒ¨
        self.message_area.verticalScrollBar().setValue(self.message_area.verticalScrollBar().maximum())
        
    # --- 7. QSS åŠ è½½ ---
    def load_stylesheet(self, filepath):
        """ä»æ–‡ä»¶åŠ è½½ QSS æ ·å¼"""
        try:
            with open(filepath, "r", encoding="utf-8") as f:
                self.setStyleSheet(f.read())
        except FileNotFoundError:
            print(f"è­¦å‘Šï¼šæ‰¾ä¸åˆ°æ ·å¼æ–‡ä»¶ {filepath}ï¼Œå°†ä½¿ç”¨é»˜è®¤æ ·å¼ã€‚")
    
    def switch_theme(self, theme_name):
        """åŠ¨æ€åˆ‡æ¢ UI ä¸»é¢˜"""
        theme_map = {
            'ios': 'qss/ios_style.qss',
            'dark': 'qss/dark_style.qss',  # ä¹‹å‰çš„æ·±è‰²ä¸»é¢˜
            'synthwave': 'qss/synthwave_style.qss', # æ–°çš„ Synthwave ä¸»é¢˜
            # TODO: åç»­å¯æ·»åŠ  'lofi', 'ambient' ç­‰ä¸»é¢˜
        }
        
        filename = theme_map.get(theme_name, 'ios_style.qss') # æ‰¾ä¸åˆ°åˆ™å›é€€
        self.load_stylesheet(filename)

# --- 8. åº”ç”¨å¯åŠ¨ ---
if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = FlowRadioApp()
    
    # ç¤ºä¾‹ï¼šå¯åŠ¨åæ·»åŠ å‡ æ¡æ¶ˆæ¯
    window.add_message("æ¬¢è¿æ”¶å¬ FlowRadioï¼æˆ‘æ˜¯ DJ Astroã€‚", is_user=False)
    window.add_message("å½“å‰ç¯å¢ƒï¼šç¼–ç¨‹ä¸­ã€‚ä¸ºæ‚¨è‡ªåŠ¨æ¨è Lo-Fi éŸ³ä¹ã€‚", is_user=False)
    
    window.show()
    sys.exit(app.exec())